src/server/session.ts

```diff
@@ -119,13 +119,158 @@ namespace ts.server {
+        const sendInterval = 1000 * 60 * 5; // 5 minutes
+        let nextSendTimeMs = Date.now() + sendInterval;
+
+        const settingNames = [
+            "allowNonTsExtensions",
+            "declaration",
+            "emitBOM",
+            "emitDecoratorMetadata",
+            "experimentalAsyncFunctions",
+            "jsx",
+            "isolatedModules",
+            "module",
+            "moduleResolution",
+            "noEmit",
+            "noEmitOnError",
+            "noImplicitAny",
+            "noLib",
+            "noResolve",
+            "out",
+            "outFile",           
+            "preserveConstEnums",
+            "removeComments",
+            "suppressImplicitAnyIndexErrors",
+            "target"
+        ];
+
+        // Keeps a count of total invocations of various language service operations (rename, gotodef, etc)
+        export function countEvent(eventName: string, projectSvc: ProjectService, host: ts.System) {
+            let opts = projectSvc.getFormatCodeOptions();
+            if (opts.SendMetrics) {
+                eventCounts[eventName] = (eventCounts[eventName] || 0) + 1;
+
+                if (Date.now() > nextSendTimeMs) {
+                    registerSettings(projectSvc);                    
+                    send(host, opts.TelemetryUserID);
+                }
+            }
+        }
+
+        // Keeps track of any .d.ts files used which correspond to known versions from DefinitelyTyped
+        export function countDts(file: string) {
+            for(var i in properties) {
+                if(properties[i] == file) {
+                    return;
+                }
+            }
+            properties['dtf' + ++dtsCount] = file;
+        }
+
+        function registerSettings(svc: ProjectService) {
+            properties['host'] = svc.hostConfiguration.hostInfo;
+            properties['inferredProjects'] = svc.inferredProjects.length.toString();
+            properties['configuredProjects'] = svc.configuredProjects.length.toString();
+
+            let someProject: Project = undefined;
+            if (svc.configuredProjects.length > 0) {
+                someProject = svc.configuredProjects[0];
+            } else if (svc.inferredProjects.length > 0) {
+                someProject = svc.inferredProjects[0];
+            }
+
+            let src = someProject && someProject.projectOptions && someProject.projectOptions.compilerOptions;
+            if (src) {
+                for (var i = 0, n = settingNames.length; i < n; i++) {
+                    let setting = settingNames[i];
+                    let value: any = src[settingNames[i]];
+                    // actual value of out or outFile is PII, don't record it, just record the flag was used
+                    if((setting == "out" || setting == "outFile") && value != null && value != '') {
+                        value = 1;
+                    }
+                    properties['project.' + settingNames[i]] = value;

```

Comment:
My recollection is for 2 reasons:
1) AppInsights doesn't allow you to create arbitrary properties whose values are arrays
2) Azure Stream Analytics (which moves the data from AppInsight into SQL for the website display) has a horrendous time trying to process array valued properties and avoiding creating any additional ones is valuable from that perspective